File: examples/banking-java/src/main/java/io/holixon/axon/projection/adhoc/example/model/CurrentBalanceModel.java
package io.holixon.axon.projection.adhoc.example.model;

import io.holixon.axon.projection.adhoc.example.events.BankAccountCreatedEvent;
import io.holixon.axon.projection.adhoc.example.events.MoneyDepositedEvent;
import io.holixon.axon.projection.adhoc.example.events.MoneyWithdrawnEvent;
import org.axonframework.messaging.annotation.MessageHandler;

public record CurrentBalanceModel(
  String bankAccountId,
  int currentBalanceInEuroCent
) {

  @MessageHandler
  public CurrentBalanceModel(BankAccountCreatedEvent event) {
    this(event.bankAccountId(), 0);
  }

  @MessageHandler
  public CurrentBalanceModel on(MoneyDepositedEvent event) {
    return new CurrentBalanceModel(bankAccountId, this.currentBalanceInEuroCent + event.euroInCent());
  }

  @MessageHandler
  public CurrentBalanceModel on(MoneyWithdrawnEvent event) {
    return new CurrentBalanceModel(bankAccountId, this.currentBalanceInEuroCent - event.euroInCent());
  }
}

File: examples/performance-test/src/main/kotlin/model/CurrentBalanceModel.kt
package io.holixon.axon.projection.adhoc.model

import io.holixon.axon.projection.adhoc.ModelRepository
import org.axonframework.common.caching.NoCache
import org.axonframework.eventhandling.SequenceNumber
import org.axonframework.eventhandling.Timestamp
import org.axonframework.eventsourcing.eventstore.EventStore
import org.axonframework.messaging.annotation.MessageHandler
import org.springframework.stereotype.Component
import java.time.Instant
import java.util.*

data class CurrentBalanceModel(
  val bankAccountId: UUID,
  val currentBalanceInEuroCent: Int,
  val lastModification: Instant,
  val version: Long
) {

  @MessageHandler
  constructor(evt: BankAccountCreatedEvent, @Timestamp messageTimestamp: Instant, @SequenceNumber version: Long) : this(
    bankAccountId = evt.bankAccountId,
    currentBalanceInEuroCent = 0,
    lastModification = messageTimestamp,
    version = version,
  )

  @MessageHandler
  fun on(evt: MoneyDepositedEvent, @Timestamp messageTimestamp: Instant, @SequenceNumber version: Long): CurrentBalanceModel = copy(
    currentBalanceInEuroCent = this.currentBalanceInEuroCent + evt.amountInEuroCent,
    lastModification = messageTimestamp,
    version = version,
  )

  @MessageHandler
  fun on(evt: MoneyWithdrawnEvent, @Timestamp messageTimestamp: Instant, @SequenceNumber version: Long): CurrentBalanceModel = copy(
    currentBalanceInEuroCent = this.currentBalanceInEuroCent - evt.amountInEuroCent,
    lastModification = messageTimestamp,
    version = version,
  )
}

@Component
class CurrentBalanceModelRepository(eventStore: EventStore) :
  ModelRepository<CurrentBalanceModel>(eventStore, CurrentBalanceModel::class.java, NoCache.INSTANCE)
